src/xcode/ENA/ENA/Source/Services/PPAnalyticsCollector/PPAAnalyticsSubmissionCollector.swift

```diff
@@ -0,0 +1,220 @@
+		}
+	}
+
+	private func persistHoursSinceTestResult(_ hours: Int32, for type: CoronaTestType) {
+		switch type {
+		case .pcr:
+			store.keySubmissionMetadata?.hoursSinceTestResult = hours
+		case .antigen:
+			store.antigenKeySubmissionMetadata?.hoursSinceTestResult = hours
+		}
+	}
+
+	private func setHoursSinceTestRegistration(type: CoronaTestType) {
+		guard let registrationDate = testRegistrationDate(for: type) else {
+			Log.warning("Could not log hoursSinceTestRegistration due to testRegistrationDate is nil", log: .ppa)
+			return
+		}
+
+		let diffComponents = Calendar.current.dateComponents([.hour], from: registrationDate, to: Date())
+		let hours = Int32(diffComponents.hour ?? 0)
+		persistHoursSinceTestRegistration(hours, for: type)
+	}
+
+	private func testRegistrationDate(for type: CoronaTestType) -> Date? {
+		switch type {
+		case .pcr:
+			return coronaTestService.pcrTest?.registrationDate
+		case .antigen:
+			return coronaTestService.antigenTest?.registrationDate
+		}
+	}
+
+	private func persistHoursSinceTestRegistration(_ hours: Int32, for type: CoronaTestType) {
+		switch type {
+		case .pcr:
+			store.keySubmissionMetadata?.hoursSinceTestRegistration = hours
+		case .antigen:
+			store.antigenKeySubmissionMetadata?.hoursSinceTestRegistration = hours
+		}
+	}
+
+	private func setDaysSinceMostRecentDateAtRiskLevelAtTestRegistration(type: CoronaTestType) {
+		guard let registrationDate = testRegistrationDate(for: type) else {
+			store.keySubmissionMetadata?.daysSinceMostRecentDateAtRiskLevelAtTestRegistration = -1
+			return
+		}
+		if let mostRecentRiskCalculationDate = store.enfRiskCalculationResult?.mostRecentDateWithCurrentRiskLevel {
+			let daysSinceMostRecentDateAtRiskLevelAtTestRegistration = Calendar.utcCalendar.dateComponents([.day], from: mostRecentRiskCalculationDate, to: registrationDate).day
+			let days = Int32(daysSinceMostRecentDateAtRiskLevelAtTestRegistration ?? -1)
+			persistDaysSinceMostRecentDateAtRiskLevelAtTestRegistration(days, for: type)
+		} else {
+			persistDaysSinceMostRecentDateAtRiskLevelAtTestRegistration(-1, for: type)
+		}
+	}
+
+	private func persistDaysSinceMostRecentDateAtRiskLevelAtTestRegistration(_ days: Int32, for type: CoronaTestType) {
+		switch type {
+		case .pcr:
+			store.keySubmissionMetadata?.daysSinceMostRecentDateAtRiskLevelAtTestRegistration = days
+		case .antigen:
+			store.antigenKeySubmissionMetadata?.daysSinceMostRecentDateAtRiskLevelAtTestRegistration = days
+		}
+	}
+
+	private func setHoursSinceHighRiskWarningAtTestRegistration(type: CoronaTestType) {
+		guard let riskLevel = store.enfRiskCalculationResult?.riskLevel  else {
+			Log.warning("Could not log hoursSinceHighRiskWarningAtTestRegistration due to riskLevel is nil", log: .ppa)
+			return
+		}
+		switch riskLevel {
+		case .high:
+			guard let timeOfRiskChangeToHigh = store.dateOfConversionToHighRisk,
+				  let registrationTime = coronaTestService.pcrTest?.registrationDate else {
```

Comment:
```suggestion
				  let registrationTime = coronaTestService.coronaTest(ofType: type)?.registrationDate else {
```

I think?

Reply:
Good catch! Fixed that yesterday after implementing tests and they failed here.